src/mmda/utils/stringify.py

```diff
@@ -0,0 +1,128 @@
+        logger.warning(f"span_group {span_group} has no overlapping words in document")
+        return ""
+
+    # are words cut in half? suppose word is (0, 3), but the query asks for (2, 3). Should it include the whole word?
+    match_any_word_start = any(
+        [word.start == span_group.start for word in matched_words]
+    )
+    match_any_word_end = any([word.end == span_group.end for word in matched_words])
+    if allow_partial_word_match:
+        if not match_any_word_start:
+            logger.warning(
+                f"span_group {span_group}'s start index doesnt match w/ start of any word. output string may thus include more text than expected."
+            )
+        if not match_any_word_end:
+            logger.warning(
+                f"span_group {span_group}'s end index doesnt match w/ end of any word. output string may thus include more text than expected."
+            )
+    else:
+        if not match_any_word_start:
+            raise ValueError(
+                f"span_group {span_group}'s start index doesnt match w/ start of any word"
+            )
+        if not match_any_word_end:
+            raise ValueError(
+                f"span_group {span_group}'s end index doesnt match w/ end of any word"
+            )
+
+    # if query has disjoint spans, what should we do with the in-between symbols when stringifying?
+    if Span.are_disjoint(spans=span_group.spans):
+        if not allow_disjoint_spans:
+            raise ValueError(
+                f"span_group {span_group} has disjoint spans but allow_disjoint_spans is False"
+            )
+        else:
+            logger.warning(f"span_group {span_group} has disjoint spans")
+            if include_symbols_between_disjoint_spans:
+                # TODO: easiest is probably to convert disjoint spans into a single longer span
+                raise NotImplementedError
+
+    # if matched words are disjoint, what should we do with the in-between symbols when stringifying?
+    if Span.are_disjoint(spans=[span for word in matched_words for span in word.spans]):
+        if not allow_disjoint_spans:
+            raise ValueError(
+                f"span_group {span_group} intersects words {matched_words} which have disjoint spans but allow_disjoint_spans is False"
+            )
+        else:
+            logger.warning(
+                f"span_group {span_group} intersects words {matched_words} which have disjoint spans"
+            )
+            if include_symbols_between_disjoint_spans:
+                # TODO: easiest is probably to convert disjoint spans into a single longer span
+                raise NotImplementedError
+
+    # TODO: actually, maybe refactor this. it doesnt matter if query spangroup is disjoint
+    #       the actual handling should just happen if it's disjoint words.
+
+    # define util function that produces text from a word
+    _stringify_word = lambda word: word.text if use_word_text else None
+    if use_word_text is False:
```

Comment:
what text would a caller want used if not word text?

Reply:
chatting w/ @soldni , there are cases (e.g. debugging) where one would want to actually defer to the original `doc.symbols [ start : end ]` string. this wasn't a critical use-case so didn't implement this & left it behind `NotImplementedError`, but wanted to call this out in implementation that it's a possible configurable choice (though default should be `word.text`)